File: SystemAudioProcessor/Sources/SystemAudioProcessor/PCMResampler.swift
import Accelerate
import Foundation

/// Integer (2x / 4x / 8x) polyphase FIR upsampler.
///
/// Design notes
/// ------------
/// * **Rate independence (44.1 kHz vs 48 kHz families).** The prototype low-pass
///   is defined relative to the *input* sample period: its normalized cut-off is
///   π/L radians per input sample (i.e. half the input Nyquist, which is exactly
///   the input's usable band). Because the taps are expressed in units of input
///   samples — not absolute Hz — the SAME coefficient set is correct for any
///   input rate: 44 100, 48 000, 88 200, 96 000, … all the way up. A 2x set maps
///   44.1k→88.2k and 48k→96k identically; an 8x set maps 44.1k→352.8k and
///   48k→384k identically. Only `(factor, filterMode)` changes the taps, never
///   the absolute rate. We therefore pre-build every `(factor, filterMode)`
///   coefficient set **once at init** and the render path only selects a
///   pre-built set by index. No coefficient generation ever happens on the audio
///   thread. The `runOutputConditioningChecks` "rate-family" cases verify this
///   by oversampling the same signal at both 44 100 and 48 000 Hz and confirming
///   the outputs match to floating-point tolerance.
/// * **No per-frame allocation.** All working memory (per-channel history and
///   scratch) is allocated at init and reused. `process(...)` takes raw
///   `UnsafePointer<Float>` buffers and writes into a caller-owned output buffer.
/// * **Block continuity.** Each channel always retains the latest 127 input
///   samples, including when a short filter is selected. Every filter reads its
///   own tail from this history, so changing tap count cannot read stale samples.
/// * **Two kernels.** `process(..., kernel:)` offers `.vDSP` (Accelerate
///   `vDSP_dotpr`, default) and `.direct` (a plain scalar dot product). Both
///   produce numerically identical output; `.direct` exists as a baseline so the
///   offline benchmark can compare throughput and so the structure is not
///   hard-dependent on Accelerate.
///
/// Live PCM uses 2×; 4× and 8× remain offline. Optional filter crossfades use the
/// same history for both filters and preallocated per-channel transition state.
final class PCMResampler {
    /// Selects the dot-product kernel used by `process`. Both are real-time safe
    /// (no allocation); `.direct` is a baseline for offline benchmarking.
    enum Kernel { case vDSP, direct }

    /// One pre-built polyphase coefficient set, stored with taps *reversed*
    /// within each phase so the per-output dot product reads both the taps and
    /// the input window with stride +1 (vDSP_dotpr friendly).
    private struct CoefficientSet {
        let factor: Int
        let mode: ResamplingFilterMode
        let tapsPerPhase: Int
        /// Flat storage: `factor * tapsPerPhase` floats, laid out as
        /// `coeffs[phase * tapsPerPhase + k]` (already reversed per phase).
        let coeffs: UnsafeMutablePointer<Float>
    }

    private let channelCount: Int
    private let maxInputFrames: Int

    /// All `(factor, filterMode)` coefficient sets, keyed for O(1) selection.
    /// `minimumPhaseExperimental` aliases onto the short linear-phase set until
    /// a real minimum-phase design lands (kept out of this iteration).
    private let sets: [Int: [ResamplingFilterMode: CoefficientSet]]

    /// Per-channel history: the last `maxTapsPerPhase - 1` input samples.
    private let history: UnsafeMutablePointer<Float>
    private let historyStride: Int          // = maxTapsPerPhase (room for taps-1 + guard)
    private let maxTapsPerPhase: Int

    private struct FilterTransition {
        var mix: Float = 0
        var target: Float = 0
        var step: Float = 0
        var remaining = 0
        var initialized = false
    }
    private let transitions: UnsafeMutablePointer<FilterTransition>

    /// Per-channel scratch that holds [history tail | new input] for a single
    /// `process` call, so the dot products read a contiguous ascending window.
    private let work: UnsafeMutablePointer<Float>
    private let workStride: Int             // = maxInputFrames + maxTapsPerPhase

    init(channels: Int, maxInputFrames: Int) {
        self.channelCount = max(1, channels)
        self.maxInputFrames = max(1, maxInputFrames)
        self.maxTapsPerPhase = ResamplingFilterMode.linearPhaseLong.tapsPerPhase

        // Build every supported coefficient set up front (off any audio thread).
        var built: [Int: [ResamplingFilterMode: CoefficientSet]] = [:]
        let factors = OutputConditioningParameters.allowedOversamplingFactors
        for factor in factors {
            var perMode: [ResamplingFilterMode: CoefficientSet] = [:]
            for mode in [ResamplingFilterMode.linearPhaseShort,
                         ResamplingFilterMode.linearPhaseLong] {
                perMode[mode] = PCMResampler.buildSet(factor: factor, mode: mode)
            }
            // minimumPhaseExperimental reuses the short set (stub, see header).
            perMode[.minimumPhaseExperimental] = perMode[.linearPhaseShort]
            built[factor] = perMode
        }
        self.sets = built

        let taps = self.maxTapsPerPhase
        self.historyStride = taps
        self.history = UnsafeMutablePointer<Float>.allocate(capacity: self.channelCount * historyStride)
        history.initialize(repeating: 0, count: self.channelCount * historyStride)
        transitions = .allocate(capacity: self.channelCount)
        transitions.initialize(repeating: FilterTransition(), count: self.channelCount)

        self.workStride = self.maxInputFrames + taps
        self.work = UnsafeMutablePointer<Float>.allocate(capacity: self.channelCount * workStride)
        work.initialize(repeating: 0, count: self.channelCount * workStride)
    }

    deinit {
        // `minimumPhaseExperimental` aliases the short set, so several CoefficientSet
        // values may share one buffer. Deallocate each distinct pointer exactly once.
        var freed = Set<UnsafeMutablePointer<Float>>()
        for perMode in sets.values {
            for set in perMode.values {
                if freed.insert(set.coeffs).inserted {
                    set.coeffs.deallocate()
                }
            }
        }
        history.deallocate()
        transitions.deinitialize(count: channelCount)
        transitions.deallocate()
        work.deallocate()
    }

    // MARK: - Render path (allocation-free)

    /// Upsample `inputFrames` samples of one channel by `factor` and write the
    /// result into `output`. Returns the number of output frames written
    /// (`inputFrames * factor`).
    ///
    /// - Parameters:
    ///   - input: `inputFrames` contiguous Float samples for `channel`.
    ///   - output: caller-owned buffer of at least `inputFrames * factor` floats.
    ///   - channel: channel index (0-based). Each channel keeps independent state.
    ///   - factor: 2, 4 or 8.
    ///   - mode: filter character.
    ///   - kernel: `.vDSP` (Accelerate, default) or `.direct` (scalar loop). Both
    ///     are real-time safe and numerically identical; `.direct` is for the
    ///     offline benchmark baseline.
    ///   - transitionFrames: Output frames used to crossfade a filter change.
    ///     Zero selects the new filter immediately. A new request during a fade
    ///     starts at the current mix, so repeated changes do not reset the blend.
    ///     Short/long have different group delay: the fade prevents a sample
    ///     discontinuity, but may briefly change level through phase cancellation.
    /// - Returns: output frame count, or 0 if the set/config is invalid.
    @discardableResult
    func process(input: UnsafePointer<Float>,
                 inputFrames: Int,
                 output: UnsafeMutablePointer<Float>,
                 channel: Int,
                 factor: Int,
                 mode: ResamplingFilterMode,
                 kernel: Kernel = .vDSP,
                 transitionFrames: Int = 0) -> Int {
        guard inputFrames > 0,
              inputFrames <= maxInputFrames,
              channel >= 0, channel < channelCount,
              let shortSet = sets[factor]?[.linearPhaseShort],
              let longSet = sets[factor]?[.linearPhaseLong] else {
            return 0
        }
        let historyCount = maxTapsPerPhase - 1
        let workBase = channel * workStride
        let histBase = channel * historyStride

        // Keep a fixed-size chronological history prefix, even for the short
        // filter. Its dot product begins further into the same prefix.
        if historyCount > 0 {
            work.advanced(by: workBase).update(from: history.advanced(by: histBase),
                                               count: historyCount)
        }
        work.advanced(by: workBase + historyCount).update(from: input, count: inputFrames)

        let outCount = inputFrames * factor
        var transition = transitions[channel]
        let target: Float = mode == .linearPhaseLong ? 1 : 0
        if !transition.initialized || transitionFrames <= 0 {
            transition = FilterTransition(mix: target, target: target,
                                          step: 0, remaining: 0, initialized: true)
        } else if transition.target != target {
            transition.target = target
            transition.remaining = transitionFrames
            transition.step = (target - transition.mix) / Float(transitionFrames)
        }
        for outputFrame in 0..<outCount {
            let j = outputFrame / factor
            let phase = outputFrame - j * factor
            let anchor = workBase + j + historyCount
            let mix = min(1, max(0, transition.mix))
            if mix == 0 {
                output[outputFrame] = dotProduct(shortSet, phase: phase,
                                                 anchor: anchor, kernel: kernel)
            } else if mix == 1 {
                output[outputFrame] = dotProduct(longSet, phase: phase,
                                                 anchor: anchor, kernel: kernel)
            } else {
                let short = dotProduct(shortSet, phase: phase, anchor: anchor, kernel: kernel)
                let long = dotProduct(longSet, phase: phase, anchor: anchor, kernel: kernel)
                output[outputFrame] = short + (long - short) * mix
            }
            if transition.remaining > 0 {
                transition.remaining -= 1
                transition.mix = transition.remaining == 0
                    ? transition.target : transition.mix + transition.step
            }
        }
        transitions[channel] = transition

        // Save the last `historyCount` input samples for the next block.
        if historyCount > 0 {
            let srcStart = workBase + inputFrames
            history.advanced(by: histBase).update(from: work.advanced(by: srcStart),
                                                  count: historyCount)
        }
        return outCount
    }

    @inline(__always)
    private func dotProduct(_ set: CoefficientSet, phase: Int,
                            anchor: Int, kernel: Kernel) -> Float {
        let taps = set.tapsPerPhase
        let coefficients = set.coeffs.advanced(by: phase * taps)
        let window = work.advanced(by: anchor - (taps - 1))
        var result: Float = 0
        switch kernel {
        case .vDSP:
            vDSP_dotpr(coefficients, 1, window, 1, &result, vDSP_Length(taps))
        case .direct:
            for k in 0..<taps { result += coefficients[k] * window[k] }
        }
        return result
    }

    /// Clear the per-channel FIR memory (call off the audio thread on a reset).
    func reset(channel: Int) {
        guard channel >= 0, channel < channelCount else { return }
        history.advanced(by: channel * historyStride)
            .update(repeating: 0, count: historyStride)
        transitions[channel] = FilterTransition()
    }

    func resetAll() {
        history.update(repeating: 0, count: channelCount * historyStride)
        transitions.update(repeating: FilterTransition(), count: channelCount)
    }

    // MARK: - Coefficient design (init-time only)

    private static func buildSet(factor: Int, mode: ResamplingFilterMode) -> CoefficientSet {
        let tapsPerPhase = mode.tapsPerPhase
        let length = factor * tapsPerPhase      // prototype length (multiple of factor)

        // Prototype low-pass: cut-off at π/factor in input-rate radians.
        // h[n] = sinc((n - center)/factor), Blackman-windowed. The 1/factor
        // DC-gain scaling is the sinc argument itself (sinc(x)/factor), so we do
        // not multiply by an extra scaleFactor below.
        let center = Double(length - 1) / 2.0
        var prototype = [Double](repeating: 0, count: length)
        for n in 0..<length {
            let x = (Double(n) - center) / Double(factor)
            var sample: Double
            if abs(x) < 1e-12 {
                sample = 1.0
            } else {
                sample = sin(.pi * x) / (.pi * x)
            }
            // Blackman window.
            let w = 0.42
                - 0.5 * cos(2.0 * .pi * Double(n) / Double(length - 1))
                + 0.08 * cos(4.0 * .pi * Double(n) / Double(length - 1))
            prototype[n] = sample * w
        }

        // Polyphase decomposition + per-phase tap reversal so render reads both
        // buffers with stride +1. phase p taps (k=0..tapsPerPhase-1):
        //   forward: prototype[p + k*factor]
        //   reversed store: coeffs[p*tapsPerPhase + k] = prototype[p + (tapsPerPhase-1-k)*factor]
        //
        // Per-phase DC normalization: each phase's taps are scaled so they sum to
        // exactly 1. This guarantees unity DC gain and a perfectly flat DC
        // response regardless of windowing/truncation — a constant input passes
        // through unchanged (verified by the DC unity-gain check). Without it the
        // Blackman-windowed finite prototype leaves a small per-phase sum error
        // that shows up as DC ripple.
        let storage = UnsafeMutablePointer<Float>.allocate(capacity: length)
        for phase in 0..<factor {
            var phaseSum: Float = 0
            for k in 0..<tapsPerPhase {
                phaseSum += Float(prototype[phase + k * factor])
            }
            let norm = abs(phaseSum) > 1e-12 ? 1.0 / phaseSum : 1.0
            for k in 0..<tapsPerPhase {
                let protoIndex = phase + (tapsPerPhase - 1 - k) * factor
                storage[phase * tapsPerPhase + k] = Float(prototype[protoIndex]) * norm
            }
        }
        return CoefficientSet(factor: factor,
                              mode: mode,
                              tapsPerPhase: tapsPerPhase,
                              coeffs: storage)
    }
}

File: Source/Core/src/PcmResampler.cpp
// PcmResampler.cpp — portable polyphase FIR upsampler.
//
// Port of SystemAudioProcessor/Sources/SystemAudioProcessor/PCMResampler.swift.
// The reference's `.direct` kernel is the one ported: a plain scalar dot product
// whose accumulation order (`result += coeffs[k] * window[k]`, k ascending in
// Float) is reproduced exactly, so this implementation is bit-identical to the
// macOS `.direct` baseline. The `.vDSP` kernel is documented as numerically
// identical to it and is not part of the port.
//
// Coefficient design is copied from `PCMResampler.buildSet` line for line: the
// prototype is a sinc low-pass with a Blackman window, the polyphase branches are
// normalized to unit DC gain, and the taps are stored reversed within each phase.
// See the header for why the input sample rate never enters the math.
//
// Realtime contract: processChannel() allocates nothing, takes no lock, logs
// nothing and computes nothing but the dot products. The tables are built once in
// prepare() and the scratch/history buffers are allocated in the constructor.

#include "../include/Core/PcmResampler.h"

#include <algorithm>
#include <cmath>
#include <cstring>

namespace lowend {
namespace {

constexpr double kPi = 3.14159265358979323846;

// Blackman window as the reference writes it. The reference uses `length - 1` as
// the denominator, which is >= 63 here (shortest set is 2 x 32 taps), so there is
// no division by zero.
double blackman(uint32_t n, uint32_t length) {
    const double phase = 2.0 * kPi * static_cast<double>(n) / static_cast<double>(length - 1);
    return 0.42 - 0.5 * std::cos(phase) + 0.08 * std::cos(2.0 * phase);
}

} // namespace

PcmResampler::PcmResampler(uint32_t maxInputFrames)
    : history_(longTapsPerPhase, 0.0f),
      work_(std::max(1u, maxInputFrames) + longTapsPerPhase, 0.0f),
      maxInputFrames_(std::max(1u, maxInputFrames)) {}

PcmResampler::CoefficientSet PcmResampler::buildSet(uint32_t factor, uint32_t tapsPerPhase) {
    // Prototype length is a whole multiple of the factor, so every phase has
    // exactly tapsPerPhase taps.
    const uint32_t length = factor * tapsPerPhase;
    const double center = (static_cast<double>(length) - 1.0) / 2.0;

    // Cut-off at pi/factor in input-rate radians: h[n] = sinc((n - center)/factor),
    // Blackman-windowed. The 1/factor DC-gain scaling is the sinc argument itself,
    // so no extra scale factor is applied.
    std::vector<double> prototype(length, 0.0);
    for (uint32_t n = 0; n < length; ++n) {
        const double x = (static_cast<double>(n) - center) / static_cast<double>(factor);
        const double sample = (std::fabs(x) < 1e-12) ? 1.0 : std::sin(kPi * x) / (kPi * x);
        prototype[n] = sample * blackman(n, length);
    }

    // Polyphase decomposition, per-phase DC normalization and per-phase tap
    // reversal:
    //     coeffs[phase * tapsPerPhase + k] = prototype[phase + (taps - 1 - k) * factor] * norm
    // The normalization makes each branch sum to 1, so a constant input passes
    // through unchanged; the reversal lets the render loop read taps and window
    // with stride +1.
    CoefficientSet set;
    set.tapsPerPhase = tapsPerPhase;
    set.coeffs.assign(length, 0.0f);
    for (uint32_t phase = 0; phase < factor; ++phase) {
        float phaseSum = 0.0f;
        for (uint32_t k = 0; k < tapsPerPhase; ++k) {
            phaseSum += static_cast<float>(prototype[phase + k * factor]);
        }
        const float norm = (std::fabs(phaseSum) > 1e-12f) ? (1.0f / phaseSum) : 1.0f;
        for (uint32_t k = 0; k < tapsPerPhase; ++k) {
            const uint32_t protoIndex = phase + (tapsPerPhase - 1 - k) * factor;
            set.coeffs[phase * tapsPerPhase + k] =
                static_cast<float>(prototype[protoIndex]) * norm;
        }
    }
    return set;
}

bool PcmResampler::prepare(double inputSampleRate, uint32_t factor, uint32_t filterMode) {
    if (!std::isfinite(inputSampleRate) || !(inputSampleRate > 0.0)) {
        return false;
    }
    if (factor != 2 && factor != 4 && factor != 8) {
        return false;
    }
    if (filterMode > 2) {
        return false;
    }

    // A factor change replaces the whole filter, so the retained input history no
    // longer belongs to this configuration and is cleared — the same reset the
    // macOS engine performs when the live factor changes. A change between the
    // two implemented branches is a filter-only change: the history is preserved
    // and processChannel() crossfades (minimumPhase aliases short, so selecting it
    // is a filter-only change too).
    if (factor_ != factor) {
        reset();
    }

    shortSet_ = buildSet(factor, shortTapsPerPhase);
    longSet_ = buildSet(factor, longTapsPerPhase);
    factor_ = factor;
    filterMode_ = filterMode;
    prepared_ = true;
    return true;
}

uint32_t PcmResampler::processChannel(const float* input, uint32_t inputFrames, float* output) {
    if (!prepared_ || input == nullptr || output == nullptr) {
        return 0;
    }
    if (inputFrames == 0 || inputFrames > maxInputFrames_) {
        return 0;
    }

    // Everything below only indexes pre-allocated storage.
    const uint32_t historyCount = longTapsPerPhase - 1;

    // Chronological prefix [history tail | new input], so a dot product reads a
    // contiguous ascending window. The prefix is a fixed 127 samples even when the
    // short branch is selected; that branch simply starts further into it, which
    // is what keeps the history valid across a filter change.
    std::memcpy(work_.data(), history_.data(), historyCount * sizeof(float));
    std::memcpy(work_.data() + historyCount, input, inputFrames * sizeof(float));

    const uint32_t outCount = inputFrames * factor_;
    const float target = (filterMode_ == 1) ? 1.0f : 0.0f;
    if (!transitionInitialized_) {
        transitionMix_ = target;
        transitionTarget_ = target;
        transitionStep_ = 0.0f;
        transitionRemaining_ = 0;
        transitionInitialized_ = true;
    } else if (transitionTarget_ != target) {
        // A new request during a fade starts from the current mix, so repeated
        // filter changes do not restart the blend.
        transitionTarget_ = target;
        transitionRemaining_ = filterTransitionFrames;
        transitionStep_ = (target - transitionMix_) / static_cast<float>(filterTransitionFrames);
    }

    for (uint32_t outputFrame = 0; outputFrame < outCount; ++outputFrame) {
        const uint32_t j = outputFrame / factor_;
        const uint32_t phase = outputFrame - j * factor_;
        const uint32_t anchor = j + historyCount;
        const float mix = std::min(1.0f, std::max(0.0f, transitionMix_));
        if (mix == 0.0f) {
            output[outputFrame] = dot(shortSet_, phase, anchor);
        } else if (mix == 1.0f) {
            output[outputFrame] = dot(longSet_, phase, anchor);
        } else {
            const float shortValue = dot(shortSet_, phase, anchor);
            const float longValue = dot(longSet_, phase, anchor);
            output[outputFrame] = shortValue + (longValue - shortValue) * mix;
        }
        if (transitionRemaining_ > 0) {
            --transitionRemaining_;
            transitionMix_ = (transitionRemaining_ == 0)
                ? transitionTarget_
                : transitionMix_ + transitionStep_;
        }
    }

    // Keep the last 127 samples of the stream for the next block. Because the
    // scratch prefix is [history | input], those 127 samples are exactly
    // work[inputFrames .. inputFrames + 126] for every block size: the block's own
    // samples when inputFrames >= 127, and the unused tail of the previous history
    // followed by the block's samples when the block is shorter than 127.
    std::memcpy(history_.data(), work_.data() + inputFrames, historyCount * sizeof(float));
    return outCount;
}

uint32_t PcmResampler::maxOutputFrames(uint32_t inputFrames) {
    // A static query cannot know the configured factor, so it reports the worst
    // case the class can be prepared for.
    return inputFrames * maxFactor;
}

void PcmResampler::reset() {
    std::fill(history_.begin(), history_.end(), 0.0f);
    transitionMix_ = 0.0f;
    transitionTarget_ = 0.0f;
    transitionStep_ = 0.0f;
    transitionRemaining_ = 0;
    transitionInitialized_ = false;
}

uint32_t PcmResampler::latencyFrames() const {
    if (!prepared_) {
        return 0;
    }
    const uint32_t taps = (filterMode_ == 1) ? longTapsPerPhase : shortTapsPerPhase;
    return taps * factor_;
}

float PcmResampler::dot(const CoefficientSet& set, uint32_t phase, uint32_t anchor) const {
    const uint32_t taps = set.tapsPerPhase;
    const float* coefficients = set.coeffs.data() + phase * taps;
    const float* window = work_.data() + (anchor - (taps - 1));
    float result = 0.0f;
    for (uint32_t k = 0; k < taps; ++k) {
        result += coefficients[k] * window[k];
    }
    return result;
}

} // namespace lowend
